File: src/main/java/de/spries/fleetcommander/persistence/JoinCodes.java
package de.spries.fleetcommander.persistence;

import java.util.Collection;
import java.util.HashSet;
import java.util.Locale;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

import org.apache.commons.lang3.RandomStringUtils;

public enum JoinCodes {
	INSTANCE;

	private static final Integer MAX_ACTIVE_CODES = 5;
	private Map<String, Integer> gameIdPerCode;
	private Map<Integer, Collection<String>> codesPerGameId;

	private JoinCodes() {
		reset();
	}

	public synchronized String create(int gameId) throws JoinCodeLimitReachedException {
		codesPerGameId.putIfAbsent(gameId, new HashSet<>());
		Collection<String> gameCodes = codesPerGameId.get(gameId);
		if (gameCodes.size() >= MAX_ACTIVE_CODES) {
			throw new JoinCodeLimitReachedException("There are already " + MAX_ACTIVE_CODES
					+ " active codes for this game");
		}

		String code = null;
		do {
			code = RandomStringUtils.randomAlphanumeric(6).toLowerCase(Locale.ROOT);
		} while (gameCodes.contains(code) || code.contains("0") || code.contains("o"));

		gameIdPerCode.put(code, gameId);
		gameCodes.add(code);
		return code;
	}

	public synchronized int invalidate(String joinCode) throws InvalidCodeException {
		if (joinCode == null) {
			throw new InvalidCodeException("null is an invalid code");
		}
		String joinCodeLC = joinCode.toLowerCase(Locale.ROOT);
		Integer gameId = gameIdPerCode.remove(joinCodeLC);
		if (gameId == null) {
			throw new InvalidCodeException(joinCodeLC + " is an invalid code");
		}
		codesPerGameId.get(gameId).remove(joinCodeLC);
		return gameId;
	}

	public Collection<String> get(int gameId) {
		return codesPerGameId.getOrDefault(gameId, new HashSet<>());
	}

	public void invalidateAll(int gameId) {
		Collection<String> codes = get(gameId);
		for (String code : codes) {
			gameIdPerCode.remove(code);
		}
		codesPerGameId.remove(gameId);
	}

	protected void reset() {
		gameIdPerCode = new ConcurrentHashMap<>();
		codesPerGameId = new ConcurrentHashMap<>();
	}
}

File: src/main/kotlin/de/spries/fleetcommander/persistence/JoinCodes.kt
package de.spries.fleetcommander.persistence

import java.util.HashSet
import java.util.Locale
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ThreadLocalRandom
import kotlin.streams.asSequence

enum class JoinCodes {
    INSTANCE;

    var randomGenerator: (() -> (String)) = ({
        val charPool: List<Char> = (('a'..'z') + ('0'..'9')).filterNot { it == '0' || it == 'o' }
        ThreadLocalRandom.current()
                .ints(6, 0, charPool.size)
                .asSequence()
                .map(charPool::get)
                .joinToString("")
    })

    private val gameIdPerCode: MutableMap<String, Int> = ConcurrentHashMap()
    private val codesPerGameId: MutableMap<Int, MutableCollection<String>> = ConcurrentHashMap()

    @Synchronized
    @Throws(JoinCodeLimitReachedException::class)
    fun create(gameId: Int): String {
        codesPerGameId.putIfAbsent(gameId, HashSet())
        val gameCodes = codesPerGameId[gameId]!!
        if (gameCodes.size >= MAX_ACTIVE_CODES) {
            throw JoinCodeLimitReachedException("There are already " + MAX_ACTIVE_CODES
                    + " active codes for this game")
        }

        var code = randomGenerator.invoke()
        while (gameCodes.contains(code)) {
            code = randomGenerator.invoke()
        }

        gameIdPerCode[code] = gameId
        gameCodes.add(code)
        return code
    }

    @Synchronized
    @Throws(InvalidCodeException::class)
    fun invalidate(joinCode: String?): Int {
        if (joinCode == null) {
            throw InvalidCodeException("null is an invalid code")
        }
        val joinCodeLC = joinCode.toLowerCase(Locale.ROOT)
        val gameId = gameIdPerCode.remove(joinCodeLC) ?: throw InvalidCodeException("$joinCodeLC is an invalid code")
        codesPerGameId[gameId]!!.remove(joinCodeLC)
        return gameId
    }

    operator fun get(gameId: Int): Collection<String> {
        return codesPerGameId.getOrDefault(gameId, HashSet())
    }

    fun invalidateAll(gameId: Int) {
        val codes = get(gameId)
        for (code in codes) {
            gameIdPerCode.remove(code)
        }
        codesPerGameId.remove(gameId)
    }

    fun reset() {
        gameIdPerCode.clear()
        codesPerGameId.clear()
    }

    companion object {
        private const val MAX_ACTIVE_CODES = 5
    }
}
